geth/jail/cell.go

```diff
@@ -86,11 +88,39 @@ func (c *Cell) Stop() error {
 // event queue loop and schedules for immediate execution.
 // Intended to be used by any cell user that want's to run
 // async call, like callback.
-func (c *Cell) CallAsync(fn otto.Value, args ...interface{}) {
+func (c *Cell) CallAsync(fn otto.Value, args ...interface{}) error {
 	task := looptask.NewCallTask(fn, args...)
-	// Add a task to the queue.
-	c.loop.Add(task)
-	// And run the task immediately.
-	// It's a blocking operation.
-	c.loop.Ready(task)
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	go func() {
+		err := c.loop.Add(task)
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		err = c.loop.Ready(task)
+		if err != nil {
+			errChan <- err
+			return
+		}
+	}()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case err, ok := <-errChan:
+			if err != nil {
+				return err
+			}
+			if !ok {
```

Comment:
Those both condition checks seems redundant. There are already checks for non-nil error before writing to the channel and `ok` can be false only if the channel is closed, which in this code can happen only at the exit from the function. So this can be perfectly simplified to:

```go
    for {
        select {
        case err := <-errChan:
            return
```